### PR TITLE
[APINotes] Upstream the remaining API Notes fixes and tests

### PR DESCRIPTION
**Pull request by egorzhdan**
This upstreams the last bits of Clang API Notes functionality that is currently implemented in the Apple fork: https://github.com/apple/llvm-project/tree/next/clang/lib/APINotes

### CODE REVIEW

**Comment by egorzhdan** [comment]
This depends on https://github.com/llvm/llvm-project/pull/84772.

File: clang/test/APINotes/versioned.m [hunk 1]
```diff
@@ -0,0 +1,187 @@
+// RUN: rm -rf %t && mkdir -p %t
+
+// Build and check the unversioned module file.
+// RUN: %clang_cc1 -fmodules -fblocks -fimplicit-module-maps -fmodules-cache-path=%t/ModulesCache/Unversioned -fdisable-module-hash -fapinotes-modules  -fsyntax-only -I %S/Inputs/Headers -F %S/Inputs/Frameworks %s
```
Comment (compnerd) [inline clang/test/APINotes/versioned.m:4]: Extra space after `-fapinotes-modules`

File: clang/test/APINotes/yaml-convert-diags.c [hunk 2]
```diff
@@ -0,0 +1,6 @@
+// RUN: rm -rf %t
+// RUN: not %clang_cc1 -fsyntax-only -fapinotes  %s -I %S/Inputs/BrokenHeaders2 2>&1 | FileCheck %s
```
Comment (compnerd) [inline clang/test/APINotes/yaml-convert-diags.c:2]: Extra space after `-fapinotes`

File: clang/test/APINotes/yaml-parse-diags.c [hunk 3]
```diff
@@ -0,0 +1,6 @@
+// RUN: rm -rf %t
+// RUN: %clang_cc1 -fsyntax-only -fapinotes  %s -I %S/Inputs/BrokenHeaders -verify
```
Comment (compnerd) [inline clang/test/APINotes/yaml-parse-diags.c:2]: Extra space after `-fapinotes`

File: clang/lib/Sema/SemaObjCProperty.cpp [hunk 4]
```diff
@@ -638,15 +638,15 @@ ObjCPropertyDecl *Sema::CreatePropertyDecl(Scope *S,
...
   // selector names in anticipation of declaration of setter/getter methods.
   PDecl->setGetterName(GetterSel, GetterNameLoc);
   PDecl->setSetterName(SetterSel, SetterNameLoc);
   PDecl->setPropertyAttributesAsWritten(
                           makePropertyAttributesAsWritten(AttributesAsWritten));
 
+  ProcessDeclAttributes(S, PDecl, FD.D);
+
   if (Attributes & ObjCPropertyAttribute::kind_readonly)
     PDecl->setPropertyAttributes(ObjCPropertyAttribute::kind_readonly);
```
[the remark this answers is not in the document]
Comment (egorzhdan) [reply clang/lib/Sema/SemaObjCProperty.cpp:652]: We have a test for this in this PR, `properties.m` – that's how I discovered this :)

File: clang/test/APINotes/Inputs/BrokenHeaders/APINotes.apinotes [hunk 5]
```diff
@@ -0,0 +1,4 @@
+Name: SomeBrokenLib
+Functions:
+  - Name: do_something_with_pointers
+    Nu llabilityOfRet: O
```
[the remark this answers is not in the document]
Comment (egorzhdan) [reply clang/test/APINotes/Inputs/BrokenHeaders/APINotes.apinotes:4]: Alright, I added a comment

File: clang/test/APINotes/nullability.m [hunk 6]
```diff
@@ -0,0 +1,44 @@
+// RUN: rm -rf %t && mkdir -p %t
+// RUN: %clang_cc1 -fmodules  -fimplicit-module-maps -fmodules-cache-path=%t/ModulesCache -fapinotes-modules -Wno-private-module -fsyntax-only -I %S/Inputs/Headers -F %S/Inputs/Frameworks %s -verify
+
+// Test with Swift version 3.0. This should only affect the few APIs that have an entry in the 3.0 tables.
```
[the remark this answers is not in the document]
Comment (egorzhdan) [reply clang/test/APINotes/nullability.m:4]: Yes, in a sense that API Notes that apply attributes specific to Swift version 3.0 still exist in the SDK, and can be read by the compiler.

**Comment by egorzhdan** [comment]
@compnerd I think I addressed all of your comments, let me know if this is good to go!

**Pull request merged**